### PR TITLE
Include weapons the player doesn't have ammo for in deterrence calculation

### LINKED ISSUE #8799: Pirate deterrence isn't recalculated when buying/selling ammo
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Beyond the title, the problem is best described by the repro steps below.

### Steps to Reproduce

1. Open the attached pilot.
2. Observe that we have 6 Sparrow escorts, each with 4 meteor missile pods and no ammo.
3. Observe that the player info box shows "fleet: sitting duck (-0.0)".  This is currently intended behavior, because Ship::CalculateDeterrence() deliberately avoids counting weapons for which the ship has no ammo.
4. Go to the outfitter, select all the escorts, and buy a bunch of meteor missiles.
5. Look back at the player info box, and observe that it still says "fleet: sitting duck (-0.0)".  This is wrong; it should be higher than that.
6. Go to the outfitter and sell just one of the missile pods per escort.
7. Observe that the player info box now says "fleet: puny defenses (-2.6)".
8. Buy back up to 4 missile pods per escort.
9. Observe that the info box says "fleet: somewhat vulnerable (-3.0)".
10. Sell all the missiles.
11. Observe that we're _still_ at "fleet: somewhat vulnerable (-3.0)".  This is bad; it should be back to 0.
12. Reload the pilot.
13. Observe that we're now back to "fleet: sitting duck (-0.0)".  So reloading does trigger recalculation.

### Expected Behavior

I would have expected that, when buying ammo, it would immediately update the pirate deterrence values; that I wouldn't have to either reload the pilot or buy/sell weapons to make it recalculate.

(I encountered this when buying Sparrows and filling them with missiles, and consulting the info box to see how many I would need, and ... finding that the value was not updated.)

The obvious fix would be to add a recalculation trigger when buying ammo.  I don't know how easy that is.

An alternative fix would be to change the logic so that "whether you have any ammo" no longer counts.  I think this would be logical—given that pirates can't tell whether you're carrying any cargo (when deciding whether to attack you—perhaps they're out of system at the time), I don't see why they could tell whether you're carrying any ammo.  If they can see whether there's a missile loaded into the tube—well, traders who have run out of ammo are probably _extra_-motivated to put a fake missile into there.  I suspect military tactics in general favor weapon systems that conceal how much ammo they have left.

I prefer the second fix.  It _reduces_ code and I feel it has the right flavor.

### Screenshots

_No response_

### Link to save file

[Deterrence Test.txt](https://github.com/endless-sky/endless-sky/files/11571887/Deterrence.Test.txt)

### Operating System

Mac OS Ventura 13.3

### Game Source

Built from source

### Game Version

commit 842291f1b43c3bc08cba4dfd2f5ba245fd2bb84a

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by TomGoodIdea**
Fixes #8799.

## Summary
When calculating deterrence, missile launchers without proper ammo loaded will be treated the same as if they were fully functional.

## Testing Done and Save File
Tested with the savefile provided in #8799.

### CODE REVIEW

**Comment by Hurleveur** [comment]
I'm pretty sure we still need to calculate the AI cache again in this case, also calculatedetterence itself should probably be changed otherwise yur ship will still have 0 when taking off with 0 ammo with the current state of this PR i think
I could open a PR to this if it helps (or straight out a new one)

**Comment by TomGoodIdea** [comment]
> calculatedetterence itself should probably be changed otherwise yur ship will still have 0 when taking off with 0 ammo with the current state of this PR i think

I've done that already, it works fine.

> I'm pretty sure we still need to calculate the AI cache again in this case

Does it need to run for all ships or only player-owned ones (like before)?

**Comment by Hurleveur** [comment]
I hadnt seen that other part of the code being changed, looks good on that end

I think that this is working right now because all ammo have mass and thus change the aiCache, but massless ammo wouldnt work.
So I would just make it recalculate the aiCache when out of ammo it for all ships

**Review by Hurleveur — APPROVED** [review]
looks good to me

**Review by quyykk — APPROVED** [review]

**Comment by TomGoodIdea** [comment]
Wait, don't we need to calibrate the cache here **only** if the ammo outfit doesn't have any mass?

**Comment by quyykk** [comment]
Hmm not sure I understand? If the ammo runs out then the AI cache needs to know about it, whether the ammo outfit has mass or not.

**Comment by Zitchas** [comment]
I can't say that I'm fond of this. If a launcher has no ammo, then it shouldn't count for deterrence. An empty missile launcher intimidates no-one.

**Comment by TomGoodIdea** [comment]
> Hmm not sure I understand? If the ammo runs out then the AI cache needs to know about it, whether the ammo outfit has mass or not.

Yes, but `Engine::MoveShip` calls `Ship::UpdateCaches`, which then calls `ShipAICache::Recalibrate`, which calibrates AI cache if ship's mass has changed. So, if the ammo outfit has mass, the AI cache is calculated twice.

**Comment by Zitchas** [comment]
Reading through the original issue, there's a lot of talk about how the game doesn't recognize when the player has reloaded. That's a bug, sure and one that should be fixed. It's not a good justification for having launchers count as deterrence when empty. 

---

Thoughts on deterrance and pirate attraction:
- It should be based on how much cargo your type of ship has stock, not how much it has now.
- It seems reasonable that pirates could see if you have a missile hanging from the launcher ready to fire, but not see how many you have ready to load. (so, having even a single missile is enough to switch the launcher from "0 deterrence" to "full deterrence".) but also that it *only* takes a single ammo. The rest of the launcher could be empty.

From a lore perspective, we need to decide what's going on with pirate attraction. Do they have spies in the spaceport keeping an eye on ships? Do they have informants among the dockworkers loading ships? Informants among the player's crew? A network of outfit and cargo scanners embeded on stealth satellites across known space that are distributed such to cover every cubic inch of every system in human space? Obviously, some of these are more reasonable situations than others. However only the least reasonable of these options really explains the in-game behavior of having pirates show up everywhere, including when you've taken a diversion in a completely different direction from where you were headed originally. 

ex. If I take a job board rush bulk delivery job to deliver supplies to Earth, then the pirates using an informant in the spaceport can make a pretty good guess that I'll be travelling fairly straight line from here to there. There's no reason they should be showing up somewhere a half dozen jumps in the opposite direction looking for me.

----

Getting into more complexity, there's different kinds of launchers. If a launcher is entirely self-contained within the ship with nothing visible from the outside, then it shouldn't be counted for deterrence regardless of whether it has ammo or not. 

But if it has a distinctive and easily recognizeable protrusion (say, a torpedo exit port or something like that) where the presence of the launcher is easy to see but we can't see into it to see if there's ammo, then it should count regardless.

And if it's a missile style launcher where the launcher is hanging outside the ship, then it should be pretty easy to see if it's loaded or not. Loaded launchers obviously provide deterrence, unloaded ones do not.

-----

Then there's the fact that alien weaponry probably shouldn't be handled normally at all. Either "They've got alien weaponry they must be super dangerous, stay as absolutely far away as possible from them." or "Looks like some junk strapped to their ship to look like a weapon, ignore it entirely." In neither case should the pirates be getting an accurate read on how much DPS that weapon is capable of.

**Comment by Hurleveur** [comment]
Tbh I think gameplay should make lore and not the opposite, so we should ask ourselves what is the best way to have this happen gameplay wise.
Now, I dont really have the answer to that, I just have an opinion.
Howeved I can say that performance wise I prefer this approach and dont see an issue with it for gameplay much (missile launchers are not that cheap). Going from needing one stock torpedo to 0 changes nothing significant for gameplay and the launchers arent that cheap.

Also I disagree with looking a the stock model for cargo as expending your cargo capacity should have consequences, since that is why raids exist in the first place: to prevent mindless cargo spam.

As for the weapon dps they should fear alien weapons but they already do if they do damage and I have a PR to make special damage matter as well (even lore wise pirates will see the effect it has)

**Comment by Zitchas** [comment]
> Also I disagree with looking at the stock model for cargo as expending your cargo capacity should have consequences, since that is why raids exist in the first place: to prevent mindless cargo spam (you'd just make ppl modify warships to have cargo, but then again it wouldnt be as efficient so mb its fine-ish)

Hmmm... That would actually allow us to spread things out a bit so that cargo ships actually have freighter-meaningful quantities of cargo in comparison to warships that have theoretically should only have minimal cargo. It'd then give people a choice: What do you want? Efficiency in carrying bulk amounts of cargo? Or do you want to sneakily transport cargo without attracting attention? Pirates should probably almost never be attacking warships for their cargo, unless they've obviously stripped off all their weapons or something really obvious like that. 

If we had cargo *pods*, then I'd say that cargo pods should be obvious and attract pirates. But our current cargo bays look like they are 100% internal, and I don't think pirates should be universally able to find out what's inside our ships.

I rather like the idea of people sneakily converting warships into freighters in order to have tough, disguised, cargo transporters. They'll be able to take a beating compared to regular freighters, and not attract as much attention of the plundering kind, but suffer significant inefficiency and a much higher crew-per-ton cost.

**Comment by Hecter94** [comment]
Pirates aren't _scanning_ your ships since raids can happen without you ever being scanned, so they must be getting their information elsewhere. Potentially pirates disguised as civilians _visually_ look at your ships without scanning them, which would let them see the ship's model, and what weapons are equipped, but not any internals like what outfits you have or what cargo you're carrying.

Since they do know what cargo you're carrying, they must have another source of information as well, and the most likely source of that information is planetary dock workers.
It would be normal for there to be various dock workers that are willing to sell information to pirates. 
You buy a bunch of valuable cargo; a dock worker notes what you bought and what your ship(s) seem to be armed with and then sells that info to various pirates for a profit. 

Combining these two sources lets pirates know roughly the value of the cargo you're carrying and how well armed you are. Still, they don't have detailed information, so they may not know if your weapons have ammo or whether you have the energy generation to run the weapons you do have, or how good your shield generators are. Still, the basic information they've got from their informants is enough to plan a raid. 

If they know that your ships are equipped with a bunch of strange-looking alien weaponry, they probably aren't going to raid you, even if you can't use those weapons for whatever reason.

**Comment by Hurleveur** [comment]
Yeah you convinced me, taking base amount makes sense, that said warships aready can protect themselves and it could just encourage going further into all into cargo.

**Comment by Hurleveur** [comment]
Also regarding alien weapons this would account for them better https://github.com/endless-sky/endless-sky/pull/8827

**Comment by waterhouse** [comment]
> Reading through the original issue, there's a lot of talk about how the game doesn't recognize when the player has reloaded. That's a bug, sure and one that should be fixed. It's not a good justification for having launchers count as deterrence when empty.

The relevant paragraph is this one:
> An alternative fix would be to change the logic so that "whether you have any ammo" no longer counts. I think this would be logical—given that pirates can't tell whether you're carrying any cargo (when deciding whether to attack you—perhaps they're out of system at the time), I don't see why they could tell whether you're carrying any ammo. If they can see whether there's a missile loaded into the tube—well, traders who have run out of ammo are probably _extra_-motivated to put a fake missile into there. I suspect military tactics in general favor weapon systems that conceal how much ammo they have left.

Which I stand by.  Now, if we changed it so pirate attraction _did_ take into account what cargo you're actually carrying, then that would kill most of the above argument.  But, in the meantime, this PR seems like an improvement.

File: source/Ship.cpp [hunk 1]
```diff
@@ -4090,13 +4090,9 @@ void Ship::ExpendAmmo(const Weapon &weapon)
 		// A realistic fraction applicable to all cases cannot be computed, so assume 50%.
 		heat -= weapon.AmmoUsage() * .5 * ammo->Mass() * MAXIMUM_TEMPERATURE * Heat();
 		AddOutfit(ammo, -weapon.AmmoUsage());
-		// Only the player's ships make use of attraction and deterrence.
-		if(isYours && !OutfitCount(ammo) && ammo->AmmoUsage())
-		{
+		if(!OutfitCount(ammo) && ammo->AmmoUsage())
 			// Recalculate the AI to account for the loss of this weapon.
```
Comment (Amazinite) [inline source/Ship.cpp:4094]:
```suggestion
		// Recalculate the AI to account for the loss of this weapon.
		if(!OutfitCount(ammo) && ammo->AmmoUsage())
```

**Comment by Amazinite** [comment]
Insofar as pirates can't peer into your cargo hold to see how much cargo you actually have and only look at your capability to hold cargo, I think it makes sense to say that pirates also can't peer into your ship to see how much ammo you have. This idea of internally vs externally mounted launchers or looking to the source of your weaponry (e.g. is it alien in origin) would make the raid system far too complex, not only in implementation but also in displaying that information in game for the player to learn and keep track of.

Sure, this does create a situation where one can cheaply "defend" their fleet using unloaded launchers, but realize that that comes at the cost of not having any actual defenses when you come across normal pirate spawns.

**Comment by waterhouse** [comment]
> Sure, this does create a situation where one can cheaply "defend" their fleet using unloaded launchers,

As Hurleveur mentions, 1 ammo provided the same deterrence as full ammo.  So even if ammo _was_ costly, you could cheese it that way.

But even if we did require full ammo to get full deterrence... For reference, here is what I figured out when cheesing deterrence on a big cargo fleet:
- Among human weapons (not counting Nuclear Missile), Meteor Missile Pods have the best DPS per ton, while Meteor Missile Launchers are a good compromise between DPS per ton and DPS per weapon slot.
- Freighters and bigger cargo ships just don't have enough weapon slots and space to cancel out their own attraction, especially when you fill them with cargo expansions.  So you'll want cargo ships, plus combat ships stuffed with weapons to provide deterrence.
- Furies are good combat ships for the purpose.  Reasonably cheap, lots of weapon slots.  Sparrows and such are somewhat more efficient in deterrence per credit, but Furies also have 6 jumps (compared to Sparrows' 3) and are less likely to get killed in crossfire, and I settled on them.
- So I gave each Fury 2 Meteor Missile Launchers and 4 Meteor Missile Pods, which uses all 40 tons of weapon space.
- Each Meteor missile costs 500; the Launcher costs 15k and carries 30 missiles (so another 15k to fill it); the Pod costs 5k and carries 7 missiles (3.5k to fill it).
- Each Fury, in my design, ended up costing ... if I divide out depreciation, looks like 726k if I fill them with 88 Meteor missiles, and 682k if I don't.  The cost difference is 6%, between full ammo and no ammo.  (And note that the missiles don't take any space—the space is effectively reserved by the launchers.)

Although I _was_ optimizing and I _did_ put 1 missile into the Furies initially... I usually ended up filling them with missiles anyway, just to shut up the "Do you want to reload all the ammunition for your ships?" prompt.

Thus, in my experience, the cost of ammo is not a significant portion of the overall cost of deterrence.

**Pull request merged**